File: Backend/ProductService/src/models/voucher.model.js
'use strict';

const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const DOCUMENT_NAME = 'Vouchers';

const VoucherSchema = new Schema(
    {
        code: {
            type: String,
            required: true,
            unique: true,
            uppercase: true,
            trim: true,
            match: [/^[A-Z0-9]{6,12}$/, 'Mã phiếu giảm giá phải dài 6-12 ký tự và chỉ chứa chữ cái in hoa hoặc số']
        },
        discountType: {
            type: String,
            enum: ['PERCENTAGE', 'FIXED_AMOUNT'],
            required: true,
        },
        discountValue: {
            type: Number,
            required: true,
            min: [0, 'Giá trị giảm giá không được âm'],
            validate: {
                validator: function (value) {
                    if (this.discountType === 'PERCENTAGE') {
                        return value <= 100;
                    }
                    return true;
                },
                message: 'Giá trị phần trăm giảm giá không được vượt quá 100%'
            }
        },
        minOrderValue: {
            type: Number,
            default: 0,
            min: [0, 'Giá trị đơn hàng tối thiểu không được âm'],
        },
        maxDiscountAmount: {
            type: Number,
            default: 0,
            min: [0, 'Số tiền giảm tối đa không được âm'],
        },
        applicableProducts: { type: Array, default: [] },
        applicableCategories: { type: Array, default: [] },
        usageLimit: {
            type: Number,
            default: 0,
            min: [0, 'Giới hạn sử dụng không được âm'],
        },
        usedCount: {
            type: Number,
            default: 0,
            min: [0, 'Số lần đã sử dụng không được âm'],
        },
        userUsed: { type: Array, default: [] },
        startDate: { type: Date, required: true, },
        endDate: {
            type: Date,
            required: true,
            validate: {
                validator: function (value) {
                    return value > this.startDate;
                },
                message: 'Ngày kết thúc phải sau ngày bắt đầu'
            }
        },
        active: { type: Boolean, default: true, },
    },
    { minimize: false, timestamps: true }
);

VoucherSchema.pre('save', function (next) {
    const now = new Date();
    if (this.endDate < now) {
        this.isActive = false;
    }
    next();
});

const voucherModel = mongoose.models.voucher || mongoose.model(DOCUMENT_NAME, VoucherSchema);

module.exports = voucherModel;

File: mobile_app/lib/models/voucher_model.dart
class VoucherModel {
  final String? id;
  final String code;
  final String discountType;
  final double discountValue;
  final double minOrderValue;
  final double maxDiscountAmount;
  final List<String> applicableProducts;
  final List<String> applicableCategories;
  final int usageLimit;
  final int usedCount;
  final List<String> userUsed;
  final DateTime startDate;
  final DateTime endDate;
  final bool active;

  VoucherModel({
    this.id,
    required this.code,
    required this.discountType,
    required this.discountValue,
    this.minOrderValue = 0,
    this.maxDiscountAmount = 0,
    this.applicableProducts = const [],
    this.applicableCategories = const [],
    this.usageLimit = 0,
    this.usedCount = 0,
    this.userUsed = const [],
    required this.startDate,
    required this.endDate,
    this.active = true,
  });

  factory VoucherModel.fromJson(Map<String, dynamic> json) {
    return VoucherModel(
      id: json['_id'] as String?,
      code: json['code'] as String,
      discountType: json['discountType'] as String,
      discountValue: (json['discountValue'] as num).toDouble(),
      minOrderValue: (json['minOrderValue'] as num?)?.toDouble() ?? 0,
      maxDiscountAmount: (json['maxDiscountAmount'] as num?)?.toDouble() ?? 0,
      applicableProducts:
          (json['applicableProducts'] as List<dynamic>?)?.cast<String>() ?? [],
      applicableCategories:
          (json['applicableCategories'] as List<dynamic>?)?.cast<String>() ??
          [],
      usageLimit: (json['usageLimit'] as num?)?.toInt() ?? 0,
      usedCount: (json['usedCount'] as num?)?.toInt() ?? 0,
      userUsed: (json['userUsed'] as List<dynamic>?)?.cast<String>() ?? [],
      startDate: DateTime.parse(json['startDate'] as String),
      endDate: DateTime.parse(json['endDate'] as String),
      active: json['active'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
    if (id != null) '_id': id,
    'code': code,
    'discountType': discountType,
    'discountValue': discountValue,
    'minOrderValue': minOrderValue,
    'maxDiscountAmount': maxDiscountAmount,
    'applicableProducts': applicableProducts,
    'applicableCategories': applicableCategories,
    'usageLimit': usageLimit,
    'usedCount': usedCount,
    'userUsed': userUsed,
    'startDate': startDate.toIso8601String(),
    'endDate': endDate.toIso8601String(),
    'active': active,
  };

  String get formattedDiscount {
    if (discountType == 'PERCENTAGE') {
      return '${discountValue.toStringAsFixed(0)}%';
    } else {
      return '${discountValue.toStringAsFixed(0)} VNĐ'; 
    }
  }

  String get formattedMaxDiscount {
    return maxDiscountAmount > 0
        ? '${maxDiscountAmount.toStringAsFixed(0)} VNĐ'
        : 'Không giới hạn';
  }

  String get formattedEndDate {
    return '${endDate.day}/${endDate.month}/${endDate.year}';
  }

  String get formattedStartDate {
    return '${startDate.day}/${startDate.month}/${startDate.year}';
  }
}
